### DISCUSSION TITLE
Antlr 4.13.2 generating incorrect c# code

### DISCUSSION THREAD

**Discussion by RossFaneuf** [antlr/antlr4#4701] [Q&A] [answered]
Antlr  4.13.2 appears to generate incorrect c# code for both my own grammar and the sample Lucne grammar. For instance, the following appears in the generated parser:

_localctx.left = TokenStream.LT(1);

When it should be something like:

_localctx.left = _input.Lt(1);

I'm having a ittle trouble believing that this error is occurring (but it is), and wonder if I have some kind of configuration or download error. 
Some facts:
Running Windows 10, using Visual Studio. Antlr directory has:

 Directory of F:\Antlr4

09/19/2024  11:43 AM    <DIR>          .
09/19/2024  11:43 AM    <DIR>          ..
09/16/2024  11:35 AM         2,140,045 antlr-4.13.2-complete.jar
09/16/2024  10:41 PM           106,374 antlr4-parse.exe
09/16/2024  10:41 PM           106,370 antlr4.exe
09/19/2024  11:44 AM    <DIR>          Documentation
09/19/2024  11:39 AM    <DIR>          Examples
09/16/2024  04:15 PM    <DIR>          Tools
               3 File(s)      2,352,789 bytes
               5 Dir(s)  625,795,338,240 bytes free

Can someone offer help with this unexpected problem?

**Comment by kaby76** [comment, marked as answer]
We need a minimal reproducible example. I.e., the .g4 grammar, driver, *and most importantly the .csproj*. But, I suspect that you are using the wrong "Antlr4"!

`TokenStream.LT(1)` is correct. [Parser.TokenStream](https://github.com/antlr/antlr4/blob/2a7904a595479954ccfb1c78124fc3d7f5bebcb1/runtime/CSharp/src/Parser.cs#L644-L647) is the property that accesses the private backing [field _input](https://github.com/antlr/antlr4/blob/2a7904a595479954ccfb1c78124fc3d7f5bebcb1/runtime/CSharp/src/Parser.cs#L110). It is LT(int): https://github.com/antlr/antlr4/blob/2a7904a595479954ccfb1c78124fc3d7f5bebcb1/runtime/CSharp/src/ITokenStream.cs#L34 There is no Lt(int) in [file ITokenStream.cs](https://github.com/antlr/antlr4/blob/2a7904a595479954ccfb1c78124fc3d7f5bebcb1/runtime/CSharp/src/ITokenStream.cs).

It sounds like you are using the old [Antlr4cs fork](https://github.com/tunnelvisionlabs/antlr4cs) and all its associated packages: [Antlr4](https://www.nuget.org/packages/Antlr4), [Antlr4.Runtime](https://www.nuget.org/packages/Antlr4.Runtime), [Antlr4.CodeGenerator](https://www.nuget.org/packages/Antlr4.CodeGenerator). These packages are ancient, and completely incompatible with ["official Antlr4"](https://github.com/antlr/antlr4).

Instead, use [Antlr4.Runtime.Standard](https://www.nuget.org/packages/Antlr4.Runtime.Standard) and my build package [Antlr4BuildTasks](https://www.nuget.org/packages/Antlr4BuildTasks). You don't need to download Java or the Antlr .jar. See [AntlrHW](https://github.com/kaby76/AntlrHW) for a complete, simple example. Just `git clone https://github.com/kaby76/AntlrHW.git; cd AntlrHW/AntlrHW; dotnet build; ./bin/Debug/net8.0/AntlrHW.exe`. (Use the MSYS2 Bash shell.) Do not run `antlr4.exe` manually. Do not run `java -jar antlr-4.13.2-complete.jar` manually. Remove all generated code from the tool that you created by running the tool manually.

Also, *very important*: there is NO Antlr4.Runtime.Standard version 4.13.2 for C#. It was never released. You must use version 4.13.1. In addition, the Antlr4BuildTasks takes great measures to make sure there is no version skew between tool and runtime. It will generate code using 4.13.1 of the tool.

**Reply by RossFaneuf** [reply]
You are correct; I got the wrong Antlr Runtime from GitHub - need Antlr4.Runtime.Standard but originally loaded Antlr.Runtime. GitHub produced the correct 14.13.1 as you specified.
